pylxd/models/instance.py

```diff
@@ -0,0 +1,803 @@
+
+
+class InstanceState(object):
+    """A simple object for representing instance state."""
+
+    def __init__(self, **kwargs):
+        for key, value in six.iteritems(kwargs):
+            setattr(self, key, value)
+
+
+_InstanceExecuteResult = collections.namedtuple(
+    'InstanceExecuteResult',
+    ['exit_code', 'stdout', 'stderr'])
+
+
+class Instance(model.Model):
+    """An LXD Instance.
+
+    This class is not intended to be used directly, but rather to be used
+    via `Client.instance.create`.
+    """
+
+    architecture = model.Attribute()
+    config = model.Attribute()
+    created_at = model.Attribute()
+    devices = model.Attribute()
+    ephemeral = model.Attribute()
+    expanded_config = model.Attribute()
+    expanded_devices = model.Attribute()
+    name = model.Attribute(readonly=True)
+    description = model.Attribute()
+    profiles = model.Attribute()
+    status = model.Attribute(readonly=True)
+    last_used_at = model.Attribute(readonly=True)
+    location = model.Attribute(readonly=True)
+    type = model.Attribute(readonly=True)
+
+    status_code = model.Attribute(readonly=True)
+    stateful = model.Attribute(readonly=True)
+
+    snapshots = model.Manager()
+    files = model.Manager()
+
+    _endpoint = 'instances'
+
+    @property
+    def api(self):
+        return getattr(self.client.api, self._endpoint)[self.name]
+
+    class FilesManager(object):
+        """A pseudo-manager for namespacing file operations."""
+
+        def __init__(self, instance):
+            self._instance = instance
+            self._endpoint = getattr(
+                instance.client.api, instance._endpoint)[instance.name].files
+
+        def put(self, filepath, data, mode=None, uid=None, gid=None):
+            """Push a file to the instance.
+
+            This pushes a single file to the instances file system named by
+            the `filepath`.
+
+            :param filepath: The path in the instance to to store the data in.
+            :type filepath: str
+            :param data: The data to store in the file.
+            :type data: bytes or str
+            :param mode: The unit mode to store the file with.  The default of
+                None stores the file with the current mask of 0700, which is
+                the lxd default.
+            :type mode: Union[oct, int, str]
+            :param uid: The uid to use inside the instance. Default of None
+                results in 0 (root).
+            :type uid: int
+            :param gid: The gid to use inside the instance.  Default of None
+                results in 0 (root).
+            :type gid: int
+            :raises: LXDAPIException if something goes wrong
+            """
+            headers = self._resolve_headers(mode=mode, uid=uid, gid=gid)
+            response = self._endpoint.post(
+                params={'path': filepath},
+                data=data,
+                headers=headers or None)
+            if response.status_code == 200:
```

Comment:
Instace._endpoint is either instance, containers, or virutal-machines. This code is in the subclass FilesManage. FilesManager._endpoint is a reference to the actual object that should be used.

I ended up reusing the name since I couldn't think of a better one :\